pkg/test/resource_creator.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"bytes"
+	"errors"
+	"io/ioutil"
+	"strings"
+
+	y2j "github.com/ghodss/yaml"
+	yaml "gopkg.in/yaml.v2"
+	apps "k8s.io/api/apps/v1"
+	core "k8s.io/api/core/v1"
+	"k8s.io/api/rbac/v1beta1"
+	crd "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	extensions_scheme "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/scheme"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	filemode = int(0664)
+)
+
+func (ctx *TestCtx) GetNamespace() (string, error) {
+	if ctx.Namespace != "" {
+		return ctx.Namespace, nil
+	}
+	// create namespace
+	if *Global.Namespace != "" {
+		ctx.Namespace = *Global.Namespace
+	} else {
+		ctx.Namespace = ctx.GetID()
+		Global.Namespace = &ctx.Namespace
+	}
+	namespaceObj := &core.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ctx.Namespace}}
+	_, err := Global.KubeClient.CoreV1().Namespaces().Create(namespaceObj)
+	if err != nil {
+		return "", err
+	}
+	ctx.AddFinalizerFn(func() error {
+		return Global.KubeClient.CoreV1().Namespaces().Delete(ctx.Namespace, metav1.NewDeleteOptions(0))
+	})
+	return ctx.Namespace, nil
+}
+
+func (ctx *TestCtx) GetCRClient(yamlCR []byte) (*rest.RESTClient, error) {
+	if ctx.CRClient != nil {
+		return ctx.CRClient, nil
+	}
+	// a user may pass nil if they expect the CRClient to already exist
+	if yamlCR == nil {
+		return nil, errors.New("CRClient does not exist; yamlCR cannot be nil")
+	}
+	// get new RESTClient for custom resources
+	crConfig := Global.KubeConfig
+	yamlMap := make(map[interface{}]interface{})
+	err := yaml.Unmarshal(yamlCR, &yamlMap)
+	if err != nil {
+		return nil, err
+	}
+	groupVersion := strings.Split(yamlMap["apiVersion"].(string), "/")
+	crGV := schema.GroupVersion{Group: groupVersion[0], Version: groupVersion[1]}
+	crConfig.GroupVersion = &crGV
+	crConfig.APIPath = "/apis"
```

Comment:
Do we need to handle the case when the Group is `""`. I believe that means the APIPath should be `/api`.

Reply:
Custom resources should not be part of the core group (or group ""), so we shouldn't need to worry about that case. The GetCRClient function is only for custom resources.

Reply:
At the moment, that function gets called as a catch-all if it's something we did not specify in the CreateFromYAML function, so that could occur. I plan to support all resources in that function soon though.